File: Programming Assignment 4/List.c
//////////////////////////////////////
//||  Geoff Huang (ghuang6)       ||//
//||  CMPS 101 Professor Tantalo  ||//
//||  Programming Assignment 4    ||//
//////////////////////////////////////

// ============================================================================
// List.c is a doubly linked List ADT that stores integers in the nodes.
// ============================================================================

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// Structs ====================================================================

// private NodeObj type
typedef struct NodeObj
{
	int data;
	struct NodeObj *next;
	struct NodeObj *prev;
} NodeObj;

// private Node type
typedef NodeObj *Node;

// private ListObj type
typedef struct ListObj
{
	Node front;
	Node back;
	Node cursor;
	int length;
	int index;
} ListObj;


// Constructors-Destructors ===================================================

// Returns reference to new Node object. Initializes next and data fields
// Private
Node newNode(int data)
{
	Node N = malloc(sizeof(NodeObj));
	N->data = data;
	N->next = N->prev = NULL;
	return (N);
}

// Frees heap memory pointed to by *pN, sets *pN to NULL
// Private
void freeNode(Node *pN)
{
	if (pN != NULL && *pN != NULL)
	{
		free(*pN);
		*pN = NULL;
	}
}

// Returns reference to new empty List object
List newList(void)
{
	List L;
	L = malloc(sizeof(ListObj));
	L->front = L->back = NULL;
	L->cursor = NULL;
	L->length = 0;
	L->index = -1;
	return (L);
}

// Frees all heap memory associated with List *pL, and sets *pL to NULL
void freeList(List *pL)
{
	if (pL != NULL && *pL != NULL) 
	{ 
		while (!isEmpty(*pL)) 
		{ 
			deleteFront(*pL); // or deleteBack(*pL)
			// NOTE: deleteFront() frees Nodes
		}
		free(*pL);
		*pL = NULL;
	}
}


// Access Functions ===========================================================

// Returns true (1) if L is empty, otherwise returns false (0)
int isEmpty(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling isEmpty() on NULL List reference\n");
		exit(1);
	}
	return (L->length == 0);
}

// Returns the number of elements in the List
int length(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling length() on NULL List reference\n");
		exit(1);
	}
	return (L->length);
}

// If cursor is defined, returns the index of the cursor element
// Otherwise returns -1
int index(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling index() on NULL List reference\n");
		exit(1);
	}
	return L->index; // note: L->index = -1 when cursor is undefined
}

// Returns the value at the front of L
int front(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling front() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		printf("List Error: calling front() on an empty List\n");
		exit(1);
	}
	return (L->front->data);
}

// Returns the value at the back of L
int back(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling back() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		printf("List Error: calling back() on an empty List\n");
		exit(1);
	}
	return (L->back->data);
}

// Returns cursor element of L
int get(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling get() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		printf("List Error: calling get() on empty List");
		exit(1);
	}
	if (index(L) == -1)
	{
		printf("List Error: calling get() when cursor is not pointing at anything");
		exit(1);      
	}
	return (L->cursor->data);
}

// returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B)
{
	int eq = 0;
	Node N = NULL;
	Node M = NULL;

	if (A == NULL || B == NULL)
	{
		printf("List Error: calling equals() on NULL List reference\n");
		exit(1);
	}

	eq = (A->length == B->length);
	N = A->front;
	M = B->front;
	while (eq && N != NULL)
	{
		eq = (N->data == M->data);
		N = N->next;
		M = M->next;
	}
	return eq;
}


// Manipulation Procedures ====================================================

// Resets this List to its original empty state
void clear(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling clear() on NULL List reference\n");
		exit(1);
	}

	while (length(L) > 0)
	{
		deleteFront(L); // or deleteBack(L)
	}
}

// If List is non-empty, places the cursor under the front element
// Otherwise, does nothing
void moveFront(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling moveFront() on NULL List reference\n");
		exit(1);
	}

	if (length(L) > 0)
	{
		L->cursor = L->front;
		L->index = 0;
	}
}

// If List is non-empty, places the cursor under the back element
// Otherwise, does nothing
void moveBack(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling moveBack() on NULL List reference\n");
		exit(1);
	}

	if (length(L) > 0)
	{
		L->cursor = L->back;
		L->index = length(L) - 1;
	}
}

// If cursor is defined and not at front, moves cursor one step towards front
// If cursor is defined and at front, cursor becomes undefined
// If cursor is undefined, does nothing
void movePrev(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling movePrev() on NULL List reference\n");
		exit(1);
	}

	if (L->cursor != NULL)
	{
		L->cursor = L->cursor->prev;
		L->index--;
	}
}

// If cursor is defined and not at back, moves cursor one step towards back
// If cursor is defined and at back, cursor becomes undefined
// If cursor is undefined, does nothing
void moveNext(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling moveNext() on NULL List reference\n");
		exit(1);
	}

	if (L->cursor != NULL)
	{
		L->cursor = L->cursor->next;
		L->index++;
		if (index(L) > length(L) - 1)
		{
			L->index = -1;
		}
	}
}

// Insert new element into this List
// If List is non-empty, insertion takes place before front element
void prepend(List L, int data)
{
	Node N = newNode(data);

	if (L == NULL)
	{
		printf("List Error: calling prepend() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L)) 
	{ 
		L->front = L->back = N;
	}
	else
	{ 
		L->front->prev = N; 
		N->next = L->front;
		L->front = N;
	}
	if (index(L) >= 0)
	{
		L->index++;
	}

	L->length++;
}

// Insert new element into this List
// If List is non-empty, insertion takes place after back element
void append(List L, int data)
{
	Node N = newNode(data);

	if (L == NULL)
	{
		printf("List Error: calling prepend() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L)) 
	{ 
		L->front = L->back = N;
	}
	else
	{ 
		L->back->next = N; 
		N->prev = L->back;
		L->back = N; 
	}

	L->length++;
}

// Insert new element before cursor
// Pre: length() > 0, index >= 0
void insertBefore(List L, int data)
{
	if (isEmpty(L))
	{
		printf("List Error: insertBefore() called on empty List\n");
		exit(1);
	}
	if (index(L) < 0)
	{
		printf("List Error: insertBefore() called when cursor is not pointing at anything\n");
		exit(1);
	}

	Node N = newNode(data);  
	N->prev = L->cursor->prev;
	N->next = L->cursor;
	L->cursor->prev = N;
	if (L->cursor == L->front)
	{
		L->front = N;
	}
	else
	{
		N->prev->next = N;
	}

	L->index++;
	L->length++;
}

// Insert new element after cursor
// Pre: length() > 0, index >= 0
void insertAfter(List L, int data)
{
	if (isEmpty(L))
	{
		printf("List Error: insertAfter() called on empty List\n");
		exit(1);
	}
	if (index(L) < 0)
	{
		printf("List Error: insertAfter() called when cursor is not pointing at anything\n");
		exit(1);
	}

	Node N = newNode(data);        
	N->next = L->cursor->next;
	N->prev = L->cursor;
	L->cursor->next = N;
	if (L->cursor == L->back)
	{
		L->back = N;
	}
	else
	{
		N->next->prev = N;
	}

	L->length++;
}

// Deletes the front element
void deleteFront(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}

	Node N = L->front;

	// change cursor if pointing at front
	// (do "nothing" if it's not pointing at anything)
	if (index(L) <= 0)
	{
		L->cursor = NULL;
		L->index = -1;
	}
	else
	{
		L->index--;
	}

	// change front element
	if (length(L) == 1)
	{
		L->front = L->back = NULL;
	}
	else
	{
		L->front = L->front->next;
		L->front->prev = NULL;
	}

	L->length--; // decrement length
	freeNode(&N); // frees heap memory pointed at by N
}

// Deletes the back element
void deleteBack(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling deleteBack() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		printf("List Error: calling deleteBack() on an empty List\n");
		exit(1);
	}

	Node N = L->back;

	// change cursor if pointing at back element
	if (index(L) == length(L) - 1)
	{
		L->cursor = NULL;
		L->index = -1;
	}

	// change back element
	if (length(L) == 1)
	{
		L->front = L->back = NULL;
	}     
	else
	{
		L->back = L->back->prev;
		L->back->next = NULL;
	}

	L->length--; // decrement length
	freeNode(&N); // frees heap memory pointed at by N
}

// Deletes cursor element, making cursor undefined
void delete(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling delete() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		printf("List Error: calling delete() on an empty List\n");
		exit(1);
	}
	if (index(L) < 0)
	{
		printf("List Error: delete() called when cursor is not pointing at anything\n");
		exit(1);
	}

	if (L->cursor == L->front)
	{
		deleteFront(L);
	}
	else if (L->cursor == L->back)
	{
		deleteBack(L);
	}
	else
	{
		Node N = L->cursor;
		L->cursor->prev->next = L->cursor->next;
		L->cursor->next->prev = L->cursor->prev;
		L->length--;
		freeNode(&N); // frees heap memory pointed at by N
	}

	L->cursor = NULL;
	L->index = -1;
}


// Other Functions ============================================================

// Prints data elements in Q on a single line to stdout.
void printList(FILE *out, List L)
{
	if (L == NULL)
	{
		printf("List Error: calling printList() on NULL List reference\n");
		exit(1);
	}

	Node N = NULL;
	for (N = L->front; N != NULL; N = N->next)
	{
		fprintf(out, "%d ", N->data);
	}
	// freeNode(&N);
}

// Returns a new List identical to this L
List copyList(List L)
{
	List L2 = newList();

	Node N = L->front;
	while (N != NULL)
	{
		append(L2, N->data);
		N = N->next;
	}

	return L2;
}

File: Programming Assignment 3/List.java
//////////////////////////////////////
//||  Geoff Huang (ghuang6)       ||//
//||  CMPS 101 Professor Tantalo  ||//
//||  Programming Assignment 3    ||//
//////////////////////////////////////

// ===================================================================================
// List.java is a doubly linked list ADT with Objects in the data fields of the nodes.
// I simply modified my List.java from pa1 to accomodate for objects instead of ints.
// List of changes (pun intended):
// - I changed the data field in the Node class from int to Object
// - I changed the equals method in the Node class to use .equals() instead of ==
// - I changed the return type of front(), back(), get() functions from int to Object
// - I changed the parameter of the prepend(), append(), insertBefore(), and  
//   insertAfter() functions from int to Object
// ===================================================================================

class List 
{
	private class Node
	{
		// Fields
		Object data;
		Node next;
		Node prev;
	  
		// Constructor
		Node(Object data) {this.data = data; next = null;}
	  
		// overrides Object's toString() method
		public String toString() 
		{ 
			return data.toString(); 
		}
	  
		// overrides Object's equals() method
		public boolean equals(Object x)
		{
			boolean eq = false;
			Node that;
			if (x instanceof Node)
			{
				that = (Node) x;
				eq = (this.data.equals(that.data));
			}
			return eq;
		}
	}

	// Fields
	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	private int index;

	// Constructor
	// Creates a new empty list
	List() 
	{ 
		front = back = cursor = null; 
		length = 0; 
		index = -1;
	}


	// Access Functions ==============================================================

	// Returns the number of elements in this List
	int length() 
	{ 
		return length; 
	}

	// index()
	// If cursor is defined, returns the index of the cursor element 
	// Otherwise returns -1
	int index()
	{
		return index;
	}
   
	// Returns front element
	// Pre: length() > 0
	Object front()
	{
		if( length() <= 0 )
		{
			throw new RuntimeException(
			"List Error: front() called on empty List");
		}
		return front.data;
	}
   
	// Returns back element
	// Pre: length() > 0
	Object back()
	{
		if( length() <= 0 )
		{
			throw new RuntimeException(
			"List Error: back() called on empty List");
		}
		return back.data;
	}
   
	// Returns cursor element
	// Pre: length() > 0, index() >= 0
	Object get()
	{
		if( length() <= 0)
		{
			throw new RuntimeException(
			"List Error: get() called on empty List");
		}
		if (index() < 0)
		{
			throw new RuntimeException(
			"List Error: get() called when the cursor is not pointing at anything");
		}
		return cursor.data;
	}
   
	// Overrides Object's equals() method.  Returns true if x is a List storing
	// the same integer sequence as this List; false otherwise.
	public boolean equals(Object x)
	{
		boolean eq  = false;
		List L;
		Node N, M;
		
		if (x instanceof List)
		{
			L = (List) x;
			N = this.front;
			M = L.front;
			eq = (this.length == L.length);
			while (eq && N != null)
			{
				eq = N.equals(M);
				N = N.next;
				M = M.next;
			}
		}
		return eq;
	}
	

	// Manipulation procedures========================================================
	
	// Resets this list to its original empty state
	void clear()
	{
		while (length() > 0)
		{
			deleteFront();
		}
	}
   
	// If List is non-empty, places the cursor under the front element
	// Otherwise, does nothing
	void moveFront()
	{
		if (length() > 0)
		{
			cursor = front;
			index = 0;
		}
	}
   
	// If List is non-empty, places the cursor under the back element
	// Otherwise, does nothing
	void moveBack()
	{
		if (length() > 0)
		{
			cursor = back;
			index = length() - 1;
		}
	}
   
	// If cursor is defined and not at front, moves cursor one step towards front
	// If cursor is defined and at front, cursor becomes undefined
	// If cursor is undefined, does nothing
	void movePrev()
	{
		if (cursor != null)
		{
			cursor = cursor.prev;
			index--;
		}
	}

	// If cursor is defined and not at back, moves cursor one step towards back
	// If cursor is defined and at back, cursor becomes undefined
	// If cursor is undefined, does nothing
	void moveNext()
	{
		if (cursor != null)
		{
			cursor = cursor.next;
			index++;
			if (index() > length() - 1)
			{
				index = -1;
			}
		}
	}
   
	// Insert new element into this List
	// If List is non-empty, insertion takes place before front element
	void prepend(Object data)
	{
		Node N = new Node(data);
		if (length() == 0) 
		{ 
			front = back = N;
		}
		else
		{ 
			front.prev = N; 
			N.next = front;
			front = N; 
		}
		if (index() >= 0)
		{
			index++;
		}
			length++;
	}

	// Insert new element into this List
	// If List is non-empty, insertion takes place after back element
	void append(Object data)
	{
		Node N = new Node(data);
		if (length() == 0) 
		{ 
			front = back = N;
		}
		else
		{ 
			back.next = N; 
			N.prev = back;
			back = N; 
		}
		length++;
	}
	  
	// Insert new element before cursor
	// Pre: length() > 0, index >= 0
	void insertBefore(Object data)
	{
		if (length() <= 0)
		{
			throw new RuntimeException(
			"List Error: insertBefore() called on empty List");
		}
		if (index() < 0)
		{
			throw new RuntimeException(
			"List Error: insertBefore() called when the cursor is not pointing at anything");
		}
	  
		Node N = new Node(data);  
		N.prev = cursor.prev;
		
		N.next = cursor;
		cursor.prev = N;
			   
		index++;
		length++;
		
		if (cursor.equals(front))
		{
			front = N;
		}
		else
		{
			N.prev.next = N;
		}
	}
	  
	// Insert new element after cursor
	// Pre: length() > 0, index >= 0
	void insertAfter(Object data)
	{
		if (length() <= 0)
		{
			throw new RuntimeException(
			"List Error: insertAfter() called on empty List");
		}
		if (index() < 0)
		{
			throw new RuntimeException(
			"List Error: insertAfter() called when the cursor is not pointing at anything");
		}
	  
		Node N = new Node(data);        
		N.next = cursor.next;

		N.prev = cursor;
		cursor.next = N;
		
		length++;
		
		if (cursor.equals(back))
		{
			back = N;
		}
		else
		{
			N.next.prev = N;
		}
	}
	 
	// Deletes the front element
	// Pre: length() > 0
	void deleteFront()
	{
		if (length() <= 0)
		{
			throw new RuntimeException(
			"List Error: deleteFront() called on empty List");
		}
		
		if (index() <= 0)
		{
			cursor = null;
			index = -1;
		}
		else
		{
			index--;
		}
		
		if (length() == 1)
		{
			front = back = null;
		}
		else
		{
			front = front.next;
			front.prev = null;
		}
		length--;
	}
   
	// Deletes the back element
	// Pre: length() > 0
	void deleteBack()
	{
		if (length() <= 0)
		{
			throw new RuntimeException(
			"List Error: deleteBack() called on empty List");
		}
		
		if (index() == length() - 1)
		{
			cursor = null;
			index = -1;
		}
		
		if (length() == 1)
		{
			front = back = null;
		}     
		else
		{
			back = back.prev;
			back.next = null;
		}
		length--;
	}
	
	// Deletes cursor element, making cursor undefined
	// Pre: length() > 0, index() >= 0
	void delete()
	{
		if (length() <= 0)
		{
			throw new RuntimeException(
			"List Error: delete() called on empty List");
		}
		if (index() < 0)
		{
			throw new RuntimeException(
			"List Error: delete() called when the cursor is not pointing at anything");
		}
		
		if (cursor.equals(front))
		{
			deleteFront();
		}
		else if (cursor.equals(back))
		{
			deleteBack();
		}
		else
		{
			cursor.prev.next = cursor.next;
			cursor.next.prev = cursor.prev;
			length--;
		}
		
		cursor = null;
		index = -1;
	}


	// Other functions ===============================================================

	// Overides Object's toString() method.
	public String toString()
	{
		StringBuffer sb = new StringBuffer();
		Node N = front;
		while (N != null)
		{
			sb.append(N.toString());
			// of course the TA wrote a check to make sure we don't have
			// an extra space at the end of our string :P
			if (N.next != null)
			{
				sb.append(" ");
			}
			N = N.next;
		}
		return new String(sb);
	}
}